### PR TITLE
[registrar] Increase code sharing in generated `native_to_managed_trampoline_*` functions

### PR DESCRIPTION
**Pull request by spouliot**
Several functions were 100% identical except for their signature. However
the signature were over-specialized and simplifying them allows the
(already) present code merge feature to achieve much better results.

Final size impact will vary based on the API used (both the managed and
native linker can remove some) but the more API you use the more likely
the application included duplicate code.

**Before**

`tools/mtouch/Xamarin.iOS.registrar.ios.x86_64.m` has `native_to_managed_trampoline_[1..379]`

```
find tools/ -name Xamarin.*.a | xargs ls -l
-rw-r--r--  1 poupou  staff  4143148 11 Oct 15:09 tools//mmp/Xamarin.Mac.registrar.full.a
-rw-r--r--  1 poupou  staff  4139052 11 Oct 15:09 tools//mmp/Xamarin.Mac.registrar.full.x86_64.a
-rw-r--r--  1 poupou  staff  4143160 11 Oct 15:09 tools//mmp/Xamarin.Mac.registrar.mobile.a
-rw-r--r--  1 poupou  staff  4139064 11 Oct 15:09 tools//mmp/Xamarin.Mac.registrar.mobile.x86_64.a
-rw-r--r--  1 poupou  staff  4521752 11 Oct 15:09 tools//mtouch/Xamarin.iOS.registrar.ios.i386.a
-rw-r--r--  1 poupou  staff  9240416 11 Oct 15:09 tools//mtouch/Xamarin.iOS.registrar.ios.simulator.a
-rw-r--r--  1 poupou  staff  4714336 11 Oct 15:09 tools//mtouch/Xamarin.iOS.registrar.ios.x86_64.a
```

**After**

`tools/mtouch/Xamarin.iOS.registrar.ios.x86_64.m` has `native_to_managed_trampoline_[1..132]`

```
find tools/ -name Xamarin.*.a | xargs ls -l
-rw-r--r--  1 poupou  staff  3723012 11 Oct 14:57 tools//mmp/Xamarin.Mac.registrar.full.a
-rw-r--r--  1 poupou  staff  3718916 11 Oct 14:57 tools//mmp/Xamarin.Mac.registrar.full.x86_64.a
-rw-r--r--  1 poupou  staff  3723016 11 Oct 14:57 tools//mmp/Xamarin.Mac.registrar.mobile.a
-rw-r--r--  1 poupou  staff  3718920 11 Oct 14:57 tools//mmp/Xamarin.Mac.registrar.mobile.x86_64.a
-rw-r--r--  1 poupou  staff  3995520 11 Oct 14:57 tools//mtouch/Xamarin.iOS.registrar.ios.i386.a
-rw-r--r--  1 poupou  staff  8195748 11 Oct 14:57 tools//mtouch/Xamarin.iOS.registrar.ios.simulator.a
-rw-r--r--  1 poupou  staff  4193956 11 Oct 14:57 tools//mtouch/Xamarin.iOS.registrar.ios.x86_64.a
```

### CODE REVIEW

**Comment by rolfbjarne** [comment]
Can you gist the before and after of `Xamarin.iOS.registrar.ios.x86_64.m`?

**Comment by spouliot** [comment]
@rolfbjarne here they are

* Before https://gist.github.com/spouliot/eb4da144351624e0188636c5bed3b77d
* After https://gist.github.com/spouliot/bbbe2c798a63f51391cd5287b6168632

but likely truncated, use the raw view

**Comment by spouliot** [comment]
yes, I have flagged a few other places too :)

That should help the interpreter builds (more than others) size.

File: tools/common/StaticRegistrar.cs [hunk 1]
```diff
@@ -2326,6 +2326,19 @@ string GetUniqueTrampolineName (string suggestion)
...
 		}
 
+		// Some declarations can be generalized to NSObject for its subclasses
+		// (and System.String too as we convert it into an NSString)
+		// since the generated code, except the function signature, is identical anyway
+		string ToSimpleObjCParameterType (TypeReference type, string descriptiveMethodName, List<Exception> exceptions, MemberReference inMethod)
+		{
+			var byref = type.IsByReference;
+			var t = byref ? type.GetElementType () : type;
+			if (t.Inherits ("Foundation", "NSObject") || t.Is ("System", "String"))
```
Comment (Therzok) [inline tools/common/StaticRegistrar.cs]: Shouldn't this be something like inherits INativeObject?

**Pull request merged**